ChibiOS 17.6.3 integration test branch [no-ci]

glowtape opened the pull request:
Doesn't boot yet.

glowtape committed:
Add ChibiOS 17.6.3 code

glowtape committed:
ChibiOS: Quick build fixes.

glowtape committed:
pios_mutex: Changes for ChibiOS 17.6.3

glowtape committed:
pios_queue: Changes for ChibiOS 17.6.3

glowtape committed:
pios_semaphore: Changes for ChibiOS 17.6.3

glowtape committed:
pios_thread: Changes for ChibiOS 17.6.3

glowtape committed:
pios: Updated ChibiOS config to new format.

MEMCORE is now required for MEMPOOLS.

glowtape committed:
loadable: IRQ vector got renamed in ChibiOS' vector table.

glowtape committed:
build: Make F4 build with ChibiOS 17.6.3

glowtape committed:
linker: Random crap.

glowtape committed:
revolution: Required defines for ChibiOS to build.

glowtape committed:
seppuku: Required define to make ChibiOS build.

glowtape committed:
led crap for testing.

tracernz commented:
Did you run into a symbol name collision between ChibiOS gpio and stdperiph gpio? That was one of the issues I hit when previously looking at going to ChibiOS 3 (/2017).

tracernz commented on flight/PiOS/ChibiOS/os/common/startup/ARMCMx/compilers/GCC/crt0_v7m.S:
```diff
@@ -120,7 +120,7 @@
  */
 #if !defined(CRT0_INIT_RAM_AREAS) || defined(__DOXYGEN__)
-#define CRT0_INIT_RAM_AREAS                 TRUE
+#define CRT0_INIT_RAM_AREAS                 FALSE
```
Please don't do this in ChibiOS source... define it somewhere in our sources.

mlyle replied:
I think this is chconf kinda stuff?

tracernz replied:
Yep, although I can't find that in boardinfo or fw... did we get rid of it?

mlyle replied:
PiOS/STM32/inc/chconf.h (a per-target one takes precedence but isn't necessary)

tracernz replied:
Oh yep. That's good. :+1:

glowtape replied:
OK, I'm gonna put it there. In regards to chconf.h, If I'm getting this to work, I'll probably define a bunch of macros so that PiOS compiles both on ChibiOS 2.6.6 on F3 and below and 17.6.3 on F4, and eventually F7, if we go there. SPRF3e is probably gonna blow again with new ChibiOS.

tracernz replied:
Why use a different version on F3 and below?

glowtape replied:
I'm gonna give it a try, but SPRF3e is really damn short on space in the firmware image. Chances are that it breaks compilation again. Plus testing PipXtreme and FlyingPio is kind of an issue for me. :)

tracernz commented on flight/targets/revolution/fw/Makefile:
```diff
@@ -245,7 +246,7 @@ CDEFS += -DUSE_$(BOARD)
 CDEFS += -DPIOS_INCLUDE_DEBUG_CONSOLE
 endif
-CDEFS += -DSTM32F40_41xxx
+CDEFS += -DSTM32F40_41xxx -DSTM32F405xx
```
Perhaps you can just define STM32F405xx... check CMSIS and StdPeriph to see what it does re: the legacy STM32F40_41xxx.

tracernz commented on flight/PiOS/ChibiOS/os/hal/ports/STM32/STM32F4xx/stm32_registry.h:
```diff
@@ -56,7 +56,7 @@
 #elif defined(STM32F405xx) || defined(STM32F415xx) ||                       \
       defined(STM32F407xx) || defined(STM32F417xx)
-#define STM32F40_41xxx
+// #define STM32F40_41xxx
```
Defining STM32F405xx would be _strongly_ preferable to modifying Chibi sources.

glowtape replied:
That one stems from STM32F40_41xxx being defined in the Revo makefile. Seppuku curiously doesn't. Should probably remove the one in the Makefile.

tracernz replied:
Yes

tracernz commented on flight/PiOS/ChibiOS/os/common/startup/ARMCMx/compilers/GCC/mk/startup_stm32f4xx.mk:
```diff
@@ -6,7 +6,9 @@ STARTUPASM = $(CHIBIOS)/os/common/startup/ARMCMx/compilers/GCC/crt0_v7m.S
+             $(PIOS_DEVLIB)/Libraries/CMSIS/Device/ST/STM32F4xx \
+             $(PIOSCOMMON)/Libraries/CMSIS/Include
+#             $(CHIBIOS)/os/common/ext/CMSIS/include \
+#             $(CHIBIOS)/os/common/ext/CMSIS/ST/STM32F4xx
```
This is where you strike problems with upgrading ChibiOS. Their CMSIS impl. doesn't seem compatible with StdPeriph etc. and vice-versa.

glowtape committed:
linker script: Textdata start

glowtape committed:
chibimain: Blinking LEDs

glowtape committed:
ld script, chconf: Fix data section, change heap stuff temporarily for memcore to work, enable memcore over whole heap.

glowtape committed:
chibimain: Do blink the Revo status light, once initTask started.

Which it doesn't yet, for whatever reason.

glowtape committed:
chibimain,chconf: Rig blinkenlights to SysTick_Handler

SysTick_Handler fires just fine, if you keep within the main thread, say like that while(1) loop right after ledinit(), that's currently commented out. However if initTask is attempted to create and start, it stops firing. Presumably still getting stuck in the context switching as suspected before.

glowtape commented:
OOOOOKKK...

I was including the GCC rules.mk file, assuming its necessity, which caused drama with the CORTEX_USE_FPU flag getting redefined, if I would keep using it in chconf.h like before. So I ditched it from chconf.h. That however created a wonky state where sometimes it is defined, and sometimes not. Apparently there's parts that do autoconfigure to the MCU, and some expect explicitely an USE_FPU flag set to yes. Which I didn't. The result was that the context switching assembly bit assumed them to be there, but not in the struct defined in another header. So during context switching, it was vpop-ing more registers than there were in the struct, and happened to hit the 0x55555555 stack fill and stick it into the program counter.

I've removed the offending rules.mk file, which seems unnecessary, because all necessary flags seem to get defined within our own makefiles, and things still work. Old ChibiOS apparently has one of these, too, which I found goes unused, too.

I have my Seppuku and Revo booting up just fine now. When I get time (prob. this weekend), I'm gonna start over with the port to get a clean one, also fixing previous points you both mentioned.
